k2/torch/csrc/nbest.cu

```diff
@@ -0,0 +1,118 @@
+/**
+ * Copyright      2021  Xiaomi Corporation (authors: Fangjun Kuang)
+ *
+ * See LICENSE for clarification regarding multiple authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "k2/csrc/fsa_algo.h"
+#include "k2/torch/csrc/fsa_algo.h"
+#include "k2/torch/csrc/nbest.h"
+namespace k2 {
+
+Nbest::Nbest(const FsaClass &fsa, const RaggedShape &shape)
+    : fsa(fsa), shape(shape) {
+  K2_CHECK_EQ(fsa.fsa.NumAxes(), 3) << "Expect an FsaVec";
+  K2_CHECK_EQ(shape.NumAxes(), 2) << "Expect a shape with axes [utt][path]";
+  K2_CHECK_EQ(fsa.fsa.Dim0(), shape.NumElements());
+}
+
+Nbest Nbest::FromLattice(FsaClass &lattice, int32_t num_paths,
+                         float nbest_scale /*= 0.5*/) {
+  K2_CHECK_EQ(lattice.fsa.NumAxes(), 3);
+  K2_CHECK_GT(num_paths, 1);
+
+  torch::Tensor scores = lattice.Scores();
+  torch::Tensor saved_scores = scores.clone();
+
+  scores = scores * nbest_scale;
+  lattice.SetScores(scores);
+  Nbest ans = RandomPaths(lattice, num_paths);
+  lattice.SetScores(saved_scores);
+  return ans;
+}
+
+void Nbest::Intersect(FsaClass *lattice) {
+  K2_CHECK_EQ(lattice->fsa.NumAxes(), 3);
+  Invert(&fsa);
+  // Now fsa contains word IDs as labels and aux_labels as token IDs.
+
+  fsa.Scores().zero_();  // Just in case it has scores set
+
+  K2_CHECK(lattice->HasTensorAttr("aux_labels") ||
+           lattice->HasRaggedTensorAttr("aux_labels"));
+
+  // We don't need the aux labels for this->fsa,
+  // as we are going to use the one from lattice.
+  Fsa word_fsa_with_epsilon_self_loops;
+  RemoveEpsilonAndAddSelfLoops(fsa.fsa, fsa.Properties(),
```

Comment:
I think it is a little confusing to use both raw Fsa and FsaClass, we'd better wrap the algorithms to FsaClass for consistency. Also the IntersectDevice in line 71.

Reply:
Thanks. I've changed `IntersectDevice`.

But for `RemoveEpsilonAndAddSelfLoops`, it does not output 1-d array arc maps and I don't want to make things complicated. We don't need to handle attribute propagation for `RemoveEpsilonAndAddSelfLoops` in decoding. I think it is clear from the calling arguments whether it is a `k2::Fsa` or `k2::FsaClass`.

Reply:
Em.. either is ok for me, up to you.